dbt/dbt-run.py

```diff
@@ -133,10 +146,85 @@ def run_incremental_models():
   # update incremental models (and tables if there are any)
   subprocess.run(["dbt", "run",  "--profiles-dir", ".dbt", "--exclude", "config.materialized:view"])
 
+def get_pending_doc_count():
+  with connection() as conn:
+      with conn.cursor() as cur:
+          cur.execute(f"""
+              SELECT SUM(pending)
+              FROM {os.getenv('POSTGRES_SCHEMA')}.couchdb_progress
+          """)
+          return cur.fetchone()[0]
+
+def update_batch_status(timestamp, status):
+  with connection() as conn:
+    with conn.cursor() as cur:
+      # insert new entry
+      cur.execute(
+        f"INSERT INTO {os.getenv('POSTGRES_SCHEMA')}.dbt_batch_status (timestamp, status) VALUES (%s, %s);", [timestamp, status]
+      )
+      conn.commit()
+
+def get_last_processed_timestamp():
+  with connection() as conn:
+    with conn.cursor() as cur:
+      cur.execute(f"""
+          SELECT MAX(timestamp)
+          FROM {os.getenv('POSTGRES_SCHEMA')}.dbt_batch_status
+          WHERE status = 'success'
+      """)
+      result = cur.fetchone()
+      if result and result[0]:
+        return result[0]
+      return '1970-01-01 00:00:00'
+    
+def get_max_timestamp():
+  with connection() as conn:
+    with conn.cursor() as cur:
+      cur.execute(f"""
+          SELECT MAX(saved_timestamp)
+          FROM {os.getenv('POSTGRES_SCHEMA')}.document_metadata
+      """)
+      return cur.fetchone()[0]
+   
+def run_dbt_in_batches():
```

Comment:
@andrablaj Sonar complains that this function is too complex and I noticed we set the complexity threshold at 5. Is there some documentation I can look at on why we set that as the threshold?
![image](https://github.com/user-attachments/assets/e210df24-b442-45ca-a63c-23f6b020d2c6)

Reply:
@jkuester do you have a link to the ticket or thread where the conversation about this topic happened?

Reply:
I think this is [the ticket](https://github.com/medic/cht-docs/pull/1155) where the conversation about max complexity happened.

Reply:
Thanks @andrablaj. 

> For both cognitive complexity and max params I favor radical reductions in the enforced values to bring about impactful change

I like the intent from @craig-landry to set radical numbers but I was looking into ways to refactor this and coming up with some very convoluted ways to do it I wondered in what scenarios we should ignore the Sonar limits and when they should be revised up or down.

Reply:
Wondering what tools like Copilot suggest when trying to refactor this sort of function and reduce its complexity. I know you are using Copilot; have you tried it for this scenario?

Reply:
Yes, I gave it a try but it didn't help with the two nested if statements. I'm sure we can refactor to meet the threshold but asking more generally when the thresholds can and should be adjusted.

Reply:
I would say that if we can refactor to reduce complexity, then let's refactor! This could make the code easier to understand. I am not proficient in Python to understand how tricky the refactoring here would be. There are several team members who know Python well if you need a second/third pair of eyes (e.g. @sugat009)

> asking more generally when the thresholds can and should be adjusted.

Sometimes, it's tempting to convince ourselves that a bit of complexity here and there won’t matter much, but if everyone adopts this mindset, the repo will become complex rapidly. Sonar rules can be overruled, and sometimes, there are some false positives. But most of times, its purpose is to make the contributors strive to write as clean code as possible. 

John Ousterhout: “In order to slow the growth of complexity, you must adopt a zero-tolerance philosophy”.

I am happy to hear other people's thoughts on the topic, if you want to move the conversation in a dedicated ticket or Slack thread.